### PR TITLE
[spark] disable shuffle while writing unaware-bucket table

### PR DESCRIPTION
**Pull request by leaves12138**
Disable shuffle for spark unaware-bucket table writes.

### CODE REVIEW

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala [hunk 1]
```diff
@@ -88,29 +88,32 @@ case class WriteIntoPaimonTable(table: FileStoreTable, overwrite: Boolean, data:
...
-                  write.write(new DynamicBucketRow(new SparkRow(rowType, bucketColDropped), bucket))
-              }
-              val serializer = new CommitMessageSerializer
-              write.prepareCommit().asScala.map(serializer.serialize).toIterator
-            } finally {
-              write.close()
+      if (!isUnawareBucketTable) {
+        // unaware bucket mode, we don't shuffle while writing
+        df = df.repartition(partitionCols ++ Seq(col(BUCKET_COL)): _*)
+      }
```
Comment (YannByron) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:97]: Suggest that put these `repartition` codes outside of `val commitMessages = ...`

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala [hunk 2]
```diff
@@ -35,9 +35,19 @@ trait PaimonCommand {
...
   def getTable: Table
 
+  val bucketMode: BucketMode = getTable match {
+    case fileStoreTable: FileStoreTable =>
+      fileStoreTable.bucketMode
+    case _ =>
+      BucketMode.FIXED
+  }
+
   def isDynamicBucketTable: Boolean = {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala:45]: remove this, use scala swith

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala [hunk 3]
```diff
@@ -35,9 +35,19 @@ trait PaimonCommand {
...
-    getTable.isInstanceOf[FileStoreTable] &&
-    getTable.asInstanceOf[FileStoreTable].bucketMode == BucketMode.DYNAMIC
+    bucketMode == BucketMode.DYNAMIC
+  }
+
+  def isUnawareBucketTable: Boolean = {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala:49]: remove this, use scala swith

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala [hunk 4]
```diff
@@ -93,29 +93,33 @@ case class WriteIntoPaimonTable(_table: FileStoreTable, saveMode: SaveMode, data
       withBucketCol.mapPartitions(commonBucketProcessor.processPartition)(withBucketDataEncoder)
     }
 
-    val commitMessages =
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:96]: unaware bucket dose not need to assign bucket too.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala [hunk 5]
```diff
@@ -77,45 +77,56 @@ case class WriteIntoPaimonTable(_table: FileStoreTable, saveMode: SaveMode, data
...
-                row =>
-                  val bucket = row.getInt(bucketColIdx)
-                  val bucketColDropped = originFromRow(toRow(row))
-                  write.write(new DynamicBucketRow(new SparkRow(rowType, bucketColDropped), bucket))
-              }
-              val serializer = new CommitMessageSerializer
-              write.prepareCommit().asScala.map(serializer.serialize).toIterator
-            } finally {
-              write.close()
+    def repartition(ds: Dataset[Row]) = {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:85]: repartitionByBucket

**Review by JingsongLi — COMMENTED** [review]
Maybe add a test for unaware bucket?

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala [hunk 6]
```diff
@@ -77,45 +77,56 @@ case class WriteIntoPaimonTable(_table: FileStoreTable, saveMode: SaveMode, data
...
-        withBucketCol.repartition(primaryKeyCols: _*)
-      } else {
-        withBucketCol
-      }
-      val numSparkPartitions = partitioned.rdd.getNumPartitions
-      val dynamicBucketProcessor =
-        DynamicBucketProcessor(table, rowType, bucketColIdx, numSparkPartitions, toRow, fromRow)
-      partitioned.mapPartitions(dynamicBucketProcessor.processPartition)(withBucketDataEncoder)
-    } else {
+    def commonBucketProcessed = {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:80]: `def assignFixedBucket`?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**